src/components/connectAdvanced.js

```diff
@@ -112,190 +106,100 @@ export default function connectAdvanced(
-        // instance, preventing the instance from being garbage collected, and that would be bad
-        const withExtras = { ...props }
-        if (withRef) withExtras.ref = this.setWrappedInstance
-        if (renderCountProp) withExtras[renderCountProp] = this.renderCount++
-        if (this.propsMode && this.subscription) withExtras[subscriptionKey] = this.subscription
-        return withExtras
+        if (connectOptions.pure) {
+          return <PureWrapper {...derivedProps} forwardRef={forwardRef}/>
+        }
+        return <WrappedComponent {...derivedProps} ref={forwardRef} />
       }
 
       render() {
-        const selector = this.selector
-        selector.shouldComponentUpdate = false
-
-        if (selector.error) {
-          throw selector.error
-        } else {
-          return createElement(WrappedComponent, this.addExtraProps(selector.props))
-        }
+        const MyConsumer = this.props.consumer || consumer
+        return (
+          <MyConsumer>
+            {this.renderWrappedComponent}
+          </MyConsumer>
+        )
       }
     }
 
-    /* eslint-enable react/no-deprecated */
-
     Connect.WrappedComponent = WrappedComponent
     Connect.displayName = displayName
-    Connect.childContextTypes = childContextTypes
-    Connect.contextTypes = contextTypes
-    Connect.propTypes = contextTypes
-
-    if (process.env.NODE_ENV !== 'production') {
-      Connect.prototype.componentWillUpdate = function componentWillUpdate() {
-        // We are hot reloading!
-        if (this.version !== version) {
-          this.version = version
-          this.initSelector()
-
-          // If any connected descendants don't hot reload (and resubscribe in the process), their
-          // listeners will be lost when we unsubscribe. Unfortunately, by copying over all
-          // listeners, this does mean that the old versions of connected descendants will still be
-          // notified of state changes; however, their onStateChange function is a no-op so this
-          // isn't a huge deal.
-          let oldListeners = [];
+    Connect.propTypes = {
+      consumer: propTypes.object,
+      forwardRef: propTypes.oneOfType([
+        propTypes.func,
+        propTypes.object
+      ])
+    }
 
-          if (this.subscription) {
-            oldListeners = this.subscription.listeners.get()
-            this.subscription.tryUnsubscribe()
-          }
-          this.initSubscription()
-          if (shouldHandleStateChanges) {
-            this.subscription.trySubscribe()
-            oldListeners.forEach(listener => this.subscription.listeners.subscribe(listener))
-          }
-        }
-      }
+    function forwardRef(props, ref) {
+      return <Connect {...props} forwardRef={ref} />
     }
 
-    return hoistStatics(Connect, WrappedComponent)
+    forwardRef.displayName = Connect.displayName
+    const forwarded = React.forwardRef(forwardRef)
+    forwarded.displayName = Connect.displayName
```

Comment:
Do you need to do this twice? Also, you can just use `displayName` variable that's locally scoped here.

Reply:
I removed the first one (commit pending). We need Connect's version, as it puts `Connect()` around the `wrappedComponentName`

Reply:
Connect doesn't do this, `getDisplayName` does it, which is set to the `displayName` var in this closure. It's a nitpick more than anything, so it's not a necessary change.